File: src/main/java/com/clinic/infrastructure/messaging/ServiceBusEventPublisher.java
package com.clinic.infrastructure.messaging;

import com.azure.core.credential.TokenCredential;
import com.azure.identity.DefaultAzureCredentialBuilder;
import com.azure.messaging.servicebus.ServiceBusClientBuilder;
import com.azure.messaging.servicebus.ServiceBusMessage;
import com.azure.messaging.servicebus.ServiceBusSenderClient;
import com.azure.messaging.servicebus.administration.ServiceBusAdministrationClientBuilder;
import com.clinic.domain.entities.Appointment;
import com.clinic.domain.ports.AppointmentEventPublisher;
import com.clinic.infrastructure.config.CorrelationContext;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.github.resilience4j.circuitbreaker.CircuitBreaker;
import io.github.resilience4j.retry.Retry;
import java.time.Instant;
import java.util.function.Supplier;

/**
 * Service Bus adapter implementing the event publisher port. Azure equivalent of the AWS project's
 * SNS+SQS / EventBridge publishing.
 *
 * <p>Authenticates via Managed Identity (DefaultAzureCredential) using the fully-qualified
 * namespace name (SERVICEBUS_FQNS). No connection string stored in config. Uses a topic so multiple
 * subscribers (country workers) can fan out.
 */
public class ServiceBusEventPublisher implements AppointmentEventPublisher {

  private static final ObjectMapper MAPPER = new ObjectMapper();

  private final ServiceBusSenderClient createdSender;
  private final ServiceBusSenderClient completedSender;
  private final ServiceBusSenderClient cancelledSender;
  private final String fullyQualifiedNamespace;
  private final TokenCredential credential;
  private final Retry retry;
  private final CircuitBreaker circuitBreaker;

  public ServiceBusEventPublisher(
      String fullyQualifiedNamespace,
      String createdTopic,
      String completedTopic,
      String cancelledTopic,
      Retry retry,
      CircuitBreaker circuitBreaker) {
    this.fullyQualifiedNamespace = fullyQualifiedNamespace;
    this.retry = retry;
    this.circuitBreaker = circuitBreaker;
    this.credential = new DefaultAzureCredentialBuilder().build();
    this.createdSender = buildSender(fullyQualifiedNamespace, credential, createdTopic);
    this.completedSender = buildSender(fullyQualifiedNamespace, credential, completedTopic);
    this.cancelledSender = buildSender(fullyQualifiedNamespace, credential, cancelledTopic);
  }

  /** Visible for testing — lets tests inject mocked {@link ServiceBusSenderClient}s. */
  ServiceBusEventPublisher(
      String fullyQualifiedNamespace,
      ServiceBusSenderClient createdSender,
      ServiceBusSenderClient completedSender,
      ServiceBusSenderClient cancelledSender,
      Retry retry,
      CircuitBreaker circuitBreaker) {
    this.fullyQualifiedNamespace = fullyQualifiedNamespace;
    this.credential = null;
    this.createdSender = createdSender;
    this.completedSender = completedSender;
    this.cancelledSender = cancelledSender;
    this.retry = retry;
    this.circuitBreaker = circuitBreaker;
  }

  private static ServiceBusSenderClient buildSender(
      String fullyQualifiedNamespace, TokenCredential credential, String topic) {
    return new ServiceBusClientBuilder()
        .fullyQualifiedNamespace(fullyQualifiedNamespace)
        .credential(credential)
        .sender()
        .topicName(topic)
        .buildClient();
  }

  public String ping() {
    try {
      new ServiceBusAdministrationClientBuilder()
          .endpoint("https://" + fullyQualifiedNamespace)
          .credential(credential)
          .buildClient()
          .getNamespaceProperties();
      return "UP";
    } catch (Exception e) {
      return "DOWN: " + e.getMessage();
    }
  }

  @Override
  public void publishCreated(Appointment a) {
    resilient(
        () -> {
          try {
            String correlationId = correlationId(a);
            ObjectNode node =
                MAPPER
                    .createObjectNode()
                    .put("eventType", "APPOINTMENT_CREATED")
                    .put("appointmentId", a.getAppointmentId())
                    .put("correlationId", correlationId)
                    .put("insuredId", a.getInsuredId())
                    .put("scheduleId", a.getScheduleId())
                    .put("countryISO", a.getCountryISO().name())
                    .put("occurredAt", Instant.now().toString());
            ServiceBusMessage message = new ServiceBusMessage(MAPPER.writeValueAsString(node));
            message.setSubject(a.getCountryISO().name());
            message.setCorrelationId(correlationId);
            message.getApplicationProperties().put("eventType", "APPOINTMENT_CREATED");
            message.getApplicationProperties().put("countryISO", a.getCountryISO().name());
            createdSender.sendMessage(message);
          } catch (RuntimeException e) {
            throw e;
          } catch (Exception e) {
            throw new RuntimeException("Failed to publish APPOINTMENT_CREATED event", e);
          }
        });
  }

  @Override
  public void publishCompleted(Appointment a) {
    resilient(
        () -> {
          try {
            String correlationId = correlationId(a);
            ObjectNode node =
                MAPPER
                    .createObjectNode()
                    .put("eventType", "APPOINTMENT_COMPLETED")
                    .put("appointmentId", a.getAppointmentId())
                    .put("correlationId", correlationId)
                    .put("insuredId", a.getInsuredId())
                    .put("countryISO", a.getCountryISO().name())
                    .put("occurredAt", Instant.now().toString());
            ServiceBusMessage message = new ServiceBusMessage(MAPPER.writeValueAsString(node));
            message.setSubject(a.getCountryISO().name());
            message.setCorrelationId(correlationId);
            message.getApplicationProperties().put("eventType", "APPOINTMENT_COMPLETED");
            message.getApplicationProperties().put("countryISO", a.getCountryISO().name());
            completedSender.sendMessage(message);
          } catch (RuntimeException e) {
            throw e;
          } catch (Exception e) {
            throw new RuntimeException("Failed to publish APPOINTMENT_COMPLETED event", e);
          }
        });
  }

  @Override
  public void publishCancelled(Appointment a) {
    resilient(
        () -> {
          try {
            String correlationId = correlationId(a);
            ObjectNode node =
                MAPPER
                    .createObjectNode()
                    .put("eventType", "APPOINTMENT_CANCELLED")
                    .put("appointmentId", a.getAppointmentId())
                    .put("correlationId", correlationId)
                    .put("insuredId", a.getInsuredId())
                    .put("countryISO", a.getCountryISO().name())
                    .put("occurredAt", Instant.now().toString());
            ServiceBusMessage message = new ServiceBusMessage(MAPPER.writeValueAsString(node));
            message.setSubject(a.getCountryISO().name());
            message.setCorrelationId(correlationId);
            message.getApplicationProperties().put("eventType", "APPOINTMENT_CANCELLED");
            message.getApplicationProperties().put("countryISO", a.getCountryISO().name());
            cancelledSender.sendMessage(message);
          } catch (RuntimeException e) {
            throw e;
          } catch (Exception e) {
            throw new RuntimeException("Failed to publish APPOINTMENT_CANCELLED event", e);
          }
        });
  }

  private static String correlationId(Appointment a) {
    String ctx = CorrelationContext.get();
    return (ctx != null && !ctx.isBlank()) ? ctx : a.getAppointmentId();
  }

  private <T> T resilient(Supplier<T> operation) {
    return Retry.decorateSupplier(retry, circuitBreaker.decorateSupplier(operation)).get();
  }

  private void resilient(Runnable operation) {
    resilient(
        () -> {
          operation.run();
          return null;
        });
  }
}

File: clinic/infrastructure/messaging/service_bus_event_publisher.py
"""Service Bus adapter implementing the event publisher port. Azure equivalent of the AWS
project's SNS+SQS / EventBridge publishing.

Authenticates via Managed Identity (DefaultAzureCredential) using the fully-qualified namespace
name. No connection string stored in config. Uses a topic so multiple subscribers (country
workers) can fan out.
"""

from __future__ import annotations

import json
from datetime import UTC, datetime
from typing import Any

from azure.servicebus import ServiceBusClient, ServiceBusMessage
from azure.servicebus.management import ServiceBusAdministrationClient

from clinic.domain.entities.appointment import Appointment
from clinic.infrastructure.config import correlation_context
from clinic.infrastructure.config.resilience import CircuitBreaker, with_retry


class ServiceBusEventPublisher:
    def __init__(
        self,
        fully_qualified_namespace: str,
        created_topic: str,
        completed_topic: str,
        cancelled_topic: str,
        circuit_breaker: CircuitBreaker,
        client: ServiceBusClient | None = None,
        admin_client: ServiceBusAdministrationClient | None = None,
    ) -> None:
        self._fully_qualified_namespace = fully_qualified_namespace
        self._circuit_breaker = circuit_breaker
        self._created_topic = created_topic
        self._completed_topic = completed_topic
        self._cancelled_topic = cancelled_topic

        if client is not None:
            self._client = client
        else:
            from azure.identity import DefaultAzureCredential

            self._client = ServiceBusClient(fully_qualified_namespace, DefaultAzureCredential())

        if admin_client is not None:
            self._admin_client = admin_client
        elif client is None:
            from azure.identity import DefaultAzureCredential

            self._admin_client = ServiceBusAdministrationClient(
                fully_qualified_namespace, DefaultAzureCredential()
            )
        else:
            self._admin_client = None

    def ping(self) -> str:
        try:
            if self._admin_client is None:
                raise RuntimeError("Administration client not configured")
            self._admin_client.get_namespace_properties()
            return "UP"
        except Exception as e:  # noqa: BLE001
            return f"DOWN: {e}"

    def publish_created(self, appointment: Appointment) -> None:
        self._resilient(
            lambda: self._send(
                self._created_topic, "APPOINTMENT_CREATED", appointment, include_schedule=True
            )
        )

    def publish_completed(self, appointment: Appointment) -> None:
        self._resilient(
            lambda: self._send(
                self._completed_topic, "APPOINTMENT_COMPLETED", appointment, include_schedule=False
            )
        )

    def publish_cancelled(self, appointment: Appointment) -> None:
        self._resilient(
            lambda: self._send(
                self._cancelled_topic, "APPOINTMENT_CANCELLED", appointment, include_schedule=False
            )
        )

    def _send(
        self, topic: str, event_type: str, appointment: Appointment, include_schedule: bool
    ) -> None:
        try:
            corr_id = self._correlation_id(appointment)
            body: dict[str, Any] = {
                "eventType": event_type,
                "appointmentId": appointment.appointment_id,
                "correlationId": corr_id,
                "insuredId": appointment.insured_id,
                "countryISO": appointment.country_iso.value,
                "occurredAt": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
            }
            if include_schedule:
                body["scheduleId"] = appointment.schedule_id

            message = ServiceBusMessage(json.dumps(body))
            message.subject = appointment.country_iso.value
            message.correlation_id = corr_id
            message.application_properties = {
                "eventType": event_type,
                "countryISO": appointment.country_iso.value,
            }
            with self._client.get_topic_sender(topic_name=topic) as sender:
                sender.send_messages(message)
        except Exception as e:  # noqa: BLE001
            raise RuntimeError(f"Failed to publish {event_type} event") from e

    @staticmethod
    def _correlation_id(appointment: Appointment) -> str:
        ctx = correlation_context.get_correlation_id()
        return ctx if ctx else appointment.appointment_id

    def _resilient(self, fn: Any) -> None:
        with_retry(lambda: self._circuit_breaker.execute(fn))
